src/Grpc.AspNetCore.Server/Internal/HttpContextServerCallContext.cs

```diff
@@ -74,7 +74,18 @@ protected override Metadata RequestHeadersCore
 
                     foreach (var header in HttpContext.Request.Headers)
                     {
-                        if (!header.Key.StartsWith(':'))
+                        // ASP.NET Core includes pseudo headers in the set of request headers
+                        // whereas, they are not in gRPC implementations. We will filter them
+                        // out when we construct the list of headers on the context.
+                        if (header.Key.StartsWith(':'))
+                        {
+                            continue;
+                        }
+                        else if (header.Key.EndsWith(Metadata.BinaryHeaderSuffix, StringComparison.OrdinalIgnoreCase))
```

Comment:
If headers are always normalized as lowercase then StringComparison.Oridinal is slightly faster.

Reply:
but they are not lowercased in the Request.Headers. When kestrel surfaces "Foo-Bin" I think we should interpret that as a binary header.

Reply:
Ah I see, this is going from HttpContext to Metadata. Looks good then.

Reply:
this might be an opportunity for future optimization - in http/2 all header keys are sent in lowercase, and ASP.NET core does seem to adjust casing for some header keys (the well known ones from HTTP1) but I assume it won't change casing for header keys it doesn't recognize - so we can be almost sure that all the received "-bin" suffixes will be lowercase in practice? Perhaps a fastpath can be added later.